src/ckb_chain/funding/funding_tx.rs

```diff
@@ -0,0 +1,253 @@
+    type_script: packed::Script,
+    /// CKB amount to be provided by the local party.
+    local_ckb_amount: u64,
+    /// CKB amount to be provided by the remote party.
+    remote_ckb_amount: u64,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct FundingRequest {
+    /// UDT channel info
+    pub udt_info: Option<FundingUdtInfo>,
+    /// The funding cell lock script args
+    pub funding_cell_lock_script_args: packed::Bytes,
+    /// Assets amount to be provided by the local party
+    pub local_amount: u64,
+    /// Fee to be provided by the local party
+    pub local_fee_rate: u64,
+    /// Assets amount to be provided by the remote party
+    pub remote_amount: u64,
+}
+
+// TODO: trace locked cells
+#[derive(Clone)]
+pub struct FundingContext {
+    pub secret_key: secp256k1::SecretKey,
+    pub rpc_url: String,
+    pub funding_source_lock_script: packed::Script,
+    pub funding_cell_lock_script: packed::Script,
+}
+
+#[allow(dead_code)]
+struct FundingTxBuilder {
+    funding_tx: FundingTx,
+    request: FundingRequest,
+    context: FundingContext,
+}
+
+impl TxBuilder for FundingTxBuilder {
+    fn build_base(
+        &self,
+        _cell_collector: &mut dyn CellCollector,
+        _cell_dep_resolver: &dyn CellDepResolver,
+        _header_dep_resolver: &dyn HeaderDepResolver,
+        _tx_dep_provider: &dyn TransactionDependencyProvider,
+    ) -> Result<TransactionView, TxBuilderError> {
+        let funding_cell = self
+            .build_funding_cell()
+            .map_err(|err| TxBuilderError::Other(err.into()))?;
+        let mut outputs: Vec<packed::CellOutput> = vec![funding_cell.0];
+        let mut outputs_data: Vec<packed::Bytes> = vec![funding_cell.1];
+
+        if let Some(ref tx) = self.funding_tx.tx {
+            for (i, output) in tx.outputs().into_iter().enumerate().skip(1) {
+                outputs.push(output.clone());
+                outputs_data.push(tx.outputs_data().get(i).unwrap_or_default().clone());
+            }
+        }
+
+        let builder = match self.funding_tx.tx {
+            Some(ref tx) => tx.as_advanced_builder(),
+            None => packed::Transaction::default().as_advanced_builder(),
+        };
+        let tx = builder
+            .set_outputs(outputs)
+            .set_outputs_data(outputs_data)
+            .build();
+
+        Ok(tx)
+    }
+}
+
+impl FundingTxBuilder {
+    fn build_funding_cell(&self) -> Result<(packed::CellOutput, packed::Bytes), FundingError> {
+        let remote_funded = self
+            .funding_tx
+            .tx
+            .as_ref()
+            .map(|tx| !tx.outputs().is_empty())
+            .unwrap_or(false);
+
+        match self.request.udt_info {
+            Some(ref udt_info) => {
+                let mut udt_amount = self.request.local_amount as u128;
+                let mut ckb_amount = udt_info.local_ckb_amount;
+
+                // If outputs is not empty, assume that the remote party has already funded.
```

Comment:
It would be better if we can put the comment here to line 110 where the variable `remote_funded` is defined. BTW, can we use something like `has_remote_funded_channel` (may be there is a better name) to make the meaning of this variable more clear?

Reply:
Agree